config/config.go

```diff
@@ -0,0 +1,224 @@
+	"github.com/hashicorp/consul-ecs/awsutil"
+	"github.com/hashicorp/consul-server-connection-manager/discovery"
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-rootcerts"
+)
+
+const (
+	defaultGRPCPort = 8503
+	defaultHTTPPort = 8501
+
+	consulCACertPemEnvVar = "CONSUL_CACERT_PEM"
+)
+
+func (c *Config) ConsulServerConnMgrConfig(taskMeta awsutil.ECSTaskMeta) (discovery.Config, error) {
+	cfg := discovery.Config{
+		Addresses: c.ConsulServers.Hosts,
+		GRPCPort:  c.ConsulServers.GRPCPort,
+	}
+
+	if c.ConsulServers.EnableTLS {
+		tlsConfig := &tls.Config{}
+
+		caCert := os.Getenv(consulCACertPemEnvVar)
+		if caCert != "" {
+			err := rootcerts.ConfigureTLS(tlsConfig, &rootcerts.Config{
+				CACertificate: []byte(caCert),
+			})
+			if err != nil {
+				return discovery.Config{}, err
+			}
+		} else if c.ConsulServers.CACertFile != "" {
+			err := rootcerts.ConfigureTLS(tlsConfig, &rootcerts.Config{
+				CAFile: c.ConsulServers.CACertFile,
+			})
+			if err != nil {
+				return discovery.Config{}, err
+			}
+		}
+
+		cfg.TLS = tlsConfig
+	}
+
+	if c.ConsulLogin.Enabled {
+		credentials, err := c.getDiscoveryCredentials(taskMeta)
+		if err != nil {
+			return discovery.Config{}, err
+		}
+
+		cfg.Credentials = credentials
+	}
+
+	return cfg, nil
+}
+
+func (c *Config) ClientConfig() *api.Config {
+	cfg := &api.Config{
+		Namespace: c.getNamespace(),
+		Partition: c.getPartition(),
+		Scheme:    "http",
+	}
+
+	if c.ConsulServers.EnableTLS {
+		cfg.Scheme = "https"
+		cfg.TLSConfig = api.TLSConfig{}
+
+		caCert := os.Getenv(consulCACertPemEnvVar)
+		if caCert != "" {
+			cfg.TLSConfig.CAPem = []byte(caCert)
+		} else if c.ConsulServers.CACertFile != "" {
+			cfg.TLSConfig.CAFile = c.ConsulServers.CACertFile
+		}
+
+		if !strings.HasPrefix(c.ConsulServers.Hosts, "exec=") {
+			cfg.TLSConfig.Address = c.ConsulServers.Hosts
+		}
+	}
+
+	return cfg
+}
+
+func (c *Config) getDiscoveryCredentials(taskMeta awsutil.ECSTaskMeta) (discovery.Credentials, error) {
+	cfg := discovery.Credentials{
+		Type: discovery.CredentialsTypeLogin,
+		Login: discovery.LoginCredential{
+			Datacenter: c.ConsulLogin.Datacenter,
+			Namespace:  c.getNamespace(),
+			Partition:  c.getPartition(),
+		},
+	}
+
+	authMethod := c.ConsulLogin.Method
+	if authMethod == "" {
+		authMethod = DefaultAuthMethodName
+	}
+	cfg.Login.AuthMethod = authMethod
+
+	meta := mergeMeta(
+		map[string]string{
+			"consul.hashicorp.com/task-id": taskMeta.TaskID(),
+			"consul.hashicorp.com/cluster": taskMeta.Cluster,
+		},
+		c.ConsulLogin.Meta,
+	)
+	cfg.Login.Meta = meta
+
+	bearerToken, err := c.createAWSBearerToken(taskMeta)
+	if err != nil {
+		return discovery.Credentials{}, err
+	}
+	cfg.Login.BearerToken = bearerToken
+
+	return cfg, nil
+}
+
+func (c *Config) getNamespace() string {
+	namespace := c.Service.Namespace
+	if namespace == "" && c.Gateway != nil {
+		namespace = c.Gateway.Namespace
+	}
+
+	return namespace
```

Comment:
minor: With these `getNamespace()` and `getPartition()` functions, it looks possible to accidentally use the namespace from the `service` config but the partition from the `gateway` config (if I've misconfigured things).

I'd recommend we add an `isGateway()` function that we can use here (and elsewhere)

```go
func (c *Config) IsGateway() bool {
	return c.Gateway != nil && c.Gateway.Kind != ""
}
```

And then this becomes:

```suggestion
	if c.IsGateway() {
		return c.Gateway.Namespace
	}
	return c.Service.Namespace
```